src/Marten/Schema/Identity/StringUniqueIdGeneration.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Collections.Generic;
+using Baseline;
+
+namespace Marten.Schema.Identity
+{
+    /// <summary>
+    /// StringUniqueIdGeneration Generation. More info https://firebase.googleblog.com/2015/02/the-2120-ways-to-ensure-unique_68.html
+		/// Copied from Gist https://gist.github.com/kiliman/ca1d9f4135078a6b24c5005113bbeea4
+    /// </summary>
+    public class StringUniqueIdGeneration : IIdGeneration, IIdGenerator<string>
+    {
+        public IEnumerable<Type> KeyTypes { get; } = new[] {typeof(string)};
+
+
+        public IIdGenerator<T> Build<T>(IDocumentSchema schema)
+        {
+            return this.As<IIdGenerator<T>>();
+        }
+
+        public string Assign(string existing, out bool assigned)
+        {
+            if (existing.IsEmpty())
+            {
+                assigned = true;
+								return FirebasePushIDGenerator.GeneratePushID();
+            }
+            assigned = false;
+            return existing;
+        }
+    }
+
+public class FirebasePushIDGenerator
+{
+    /**
+     * Fancy ID generator that creates 20-character string identifiers with the following properties:
+     *
+     * 1. They're based on timestamp so that they sort *after* any existing ids.
+     * 2. They contain 72-bits of random data after the timestamp so that IDs won't collide with other clients' IDs.
+     * 3. They sort *lexicographically* (so the timestamp is converted to characters that will sort properly).
+     * 4. They're monotonically increasing.  Even if you generate more than one in the same timestamp, the
+     *    latter ones will sort after the former ones.  We do this by using the previous random bits
+     *    but "incrementing" them by 1 (only in the case of a timestamp collision).
+     */
+    // Modeled after base64 web-safe chars, but ordered by ASCII.
+    const string PUSH_CHARS = "-0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz";
+
+    // Timestamp of last push, used to prevent local collisions if you push twice in one ms.
+    private static long lastPushTime = 0;
```

Comment:
The part that makes this difficult to unit test, especially across threads is the call to `DateTimeOffset.UtcNow`, since it is no longer deterministic.

You could add the following static property to the class that enables you to swap out the time function:

	// Func to replace current time getter useful for unit testing
	private static Func<long> _getUnixTimeMillisecondsFunc = new Func<long>(() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
	public static Func<long> GetUnixTimeMilliseconds
	{
		get { return _getUnixTimeMillisecondsFunc; }
		set { _getUnixTimeMillisecondsFunc = value; }
	}

Then update the call to use it:

	var now = GetUnixTimeMilliseconds();

In your unit test, you can then hard code a specific time that you can then use to determine the proper IDs are generated.

	FirebasePushIDGenerator.GetUnixTimeMilliseconds = new Func<long>(() => 1491859478805);

Once set, you should be able to generate multiple IDs and it should differ only by the last bit.

For testing multi-threading, you can set the specific time, then create 2 threads and generate the ID. It should have the same timestamp portion (first 8 chars), and the last 12 should be random per thread.

Hope this helps.